---
redirect_from:
  - "/recipes/resolute-raw/rawrecipe-resolute"
title: |-
  RESOLUTE experience Recipe
prev_page:
  url: /recipes/etox_raw/FAIRplus_Recipe_Ontology_mapping_the_eTox_dataset_scenario.html
  title: |-
    eTOX experience Recipe
next_page:
  url: /recipes/oncotrack_raw/rawRecipe-oncotrack.html
  title: |-
    ONCOTRACK experience Recipe
suffix: .md

comment: "***PROGRAMMATICALLY GENERATED, DO NOT EDIT. SEE ORIGINAL FILES IN /content***"
---

    <main class="jupyter-page">
    <div class="jb_cell">

<div class="cell border-box-sizing text_cell rendered"><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>RESOLUTE FAIRification Recipe</p>
<h2 id="RESOLUTE-metadata-description">RESOLUTE metadata description<a class="anchor-link" href="#RESOLUTE-metadata-description"> </a></h2><p>The RESOLUTE metadata are stored in Owncloud. There are two datasets in the RESOUTE project,  the transcriptomics dataset and proteomics dataset (Oct, 2019). The transcriptomics dataset includes: license file, gene TPM levels and transcript TPM levels, data release information (sample metadata, sequencing experiment details, data analysis protocols, etc.) The proteomics dataset includes: license file, protein expression levels, data release information.</p>
<p>Currently, there is no publication in the RESOLUTE project. The RESOLUTE data and metadata are distributed under the <a href="https://creativecommons.org/licenses/by/4.0/">CC BY 4.0 license</a>.</p>
<p>The data release plan of RESOLUTE transcriptomics dataset includes project description, study background, study summary and the metadata of three studies: RNA-Seq measurement of parental cell lines, mapping of RNA-Seq to reference genome and gene expression analysis in parent cell lines. The data creation, identification, annotation, quality control and sustainability are explained in each study.</p>
<h2 id="FAIR-assessment">FAIR assessment<a class="anchor-link" href="#FAIR-assessment"> </a></h2><p>The level of "FAIRness" of the RESOLUTE transcriptomics was first determined using the CMMI maturity model described <a href="https://docs.google.com/document/d/1URLfNpBYkCrICpizKZJ7NE29FddNNcoR0T0o_SQza7U/edit#heading=h.w0g0276fq5i6">in this document.</a> The results of this first assessment can be found <a href="https://docs.google.com/document/d/1Q_Su8kY9uNYfCV30jSIoWNdeV8GxA_DTGAcGOSZscQM/edit?usp=sharing">here</a>. In a follow up meeting we defined actions that could be taken to increase the level for certain indicators (see <a href="https://docs.google.com/document/d/1yYDcUvyXzYLfq9NZX23tbgIjCSOenSURjcRj61FMdzQ/edit?usp=sharing">this document</a>). These actions were defined within Github.</p>
<p>A <a href="https://docs.google.com/spreadsheets/d/11-jDoMbuxw8Nreurk7yKzk3EHJ54APAQnBl6VTKZPBk/edit#gid=1559176954">new FAIRplus maturity model</a> was set up, based on the RDA indicators. The focus was only on the indicators regarding Findability/Discoverability at that point. Before we could use this model however, the indicators were updated within the RDA initiative. We used their <a href="https://docs.google.com/spreadsheets/d/1mkjElFrTBPBH0QViODexNur0xNGhJqau0zkL4w8RRAw/edit?usp=sharing">new indicators</a>) (for findability, v0.01) and their level definition to determine the levels for both the transcriptomics and proteomics datasets. The result can be found <a href="https://docs.google.com/spreadsheets/d/1abQ5_sOmBWbxAZhQVEUxQ_ybI1yTi0t-tJAVY5J5VY8/edit?usp=sharing">here</a>.</p>
<p>In the meantime, an improved version of the RDA indicators was released (v0.02). In the results document a comparison/mapping is made between v0.02 and v0.01 for possible future reference.
We have updated the scoring for the transcriptomics data to this new version (v0.02).</p>
<ul>
<li><p>Results of the scoring (to v0.01):</p>
<ul>
<li><p>transcriptomics data set (as in SRA)</p>
<ul>
<li><p>58% for the mandatory indicators</p>
</li>
<li><p>63% for the recommended indicators</p>
</li>
<li><p>4 indicators were thought to be not applicable</p>
</li>
</ul>
</li>
<li><p>proteomics data set (as in OwnCloud)</p>
<ul>
<li><p>42% for the mandatory indicators</p>
</li>
<li><p>44% for the recommended indicators</p>
</li>
<li><p>6 indicators were thought to be not applicable</p>
</li>
</ul>
</li>
</ul>
</li>
</ul>
<p>The FAIRness level of the transcriptomics dataset for the HUH-7 cell line (as published on SRA) was also determined using the <a href="https://fairsharing.github.io/FAIR-Evaluator-FrontEnd/#!/#%2F!">FAIR evaluation services</a> (an online evaluator). The result was this dataset only passed for 3 out of the 22 indicators tested. This was mainly due to the limitations on machine readability of metadata and license details within SRA. The complete results can be viewed <a href="https://fairsharing.github.io/FAIR-Evaluator-FrontEnd/#!/evaluations/170">here</a>.</p>
<h2 id="ETL-pipeline-(transcriptomics-data-set)">ETL pipeline (transcriptomics data set)<a class="anchor-link" href="#ETL-pipeline-(transcriptomics-data-set)"> </a></h2><p>The metadata of RESOLUTE transcriptomics dataset are firstly extracted <a href="https://raw.githubusercontent.com/ebi-ait/FAIRPlus/master/RESOLUTE/transcriptomics/scripts/transcriptomics_metadata_extract.py">using yaml format</a> using key:value pairs. An alternative option is to store the metadata in a spreadsheet. However, the RESOLUTE metadata structure is complicated, which limits the applicability of spreadsheet format. Yaml format is preferred than JSON for its clarity and readability. The Yaml files can be converted to JSON format with <a href="https://pypi.org/project/PyYAML/">PyYAML</a>.</p>
<p>The RESOLUTE transcriptomics raw metadata are informative. Two principles are used in metadata extraction.The first one is to maintain as detailed metadata as possible, though some metadata is not supported in database submission. Secondly, the metadata should support reproducing the analysis.</p>
<p>The <a href="https://github.com/ebi-ait/FAIRPlus/blob/master/RESOLUTE/transcriptomics/data/RESOLUTE_transcriptomics_metadata_raw.json">metadata</a> of the RESOLUTE project was extracted from the data release description into JSON files.</p>
<p>The metadata was parsed according to transcriptomics community metadata standard <a href="http://fged.org/projects/minseqe/">MINSEQE</a>. The MINSEQE schema we used can be found <a href="https://github.com/FAIRsharing/mircat/tree/master/minseqe">here</a>. This version of <a href="https://github.com/ebi-ait/FAIRPlus/blob/master/RESOLUTE/transcriptomics/data/RESOLUTE_transcriptomics_metadata_MINSEQE.json">metadata</a> can not be submitted to EBI databases.</p>
<p>To support metadata submission, we modified the <a href="https://github.com/ebi-ait/FAIRPlus/blob/master/RESOLUTE/transcriptomics/data/samples_BSD.json">samples metadata</a> based on <a href="https://github.com/EBIBioSamples/biosamples-v4/blob/dev/webapps/core/src/main/resources/schemas/core/sample.json">EBI BioSamples JSON schema requirements</a>. BioSamples database doesn’t support nested JSON data, so we flattened the metadata before submission. To further submit study, assay, and other metadata, we used <a href="https://www.ebi.ac.uk/submission/">EBI Data Submission Portal</a> to submit <a href="https://github.com/ebi-ait/FAIRPlus/tree/master/RESOLUTE/transcriptomics/data/data_for_USI">study, assay, sequencing Run, sample data</a> together. We also checked if the new metadata compatible with the MINSEQE guidelines. Note! DSP doesn’t support external ontology annotation URLs currently, so we didn’t include ontology mapping in this version.</p>
<p>We created a <a href="https://github.com/ebi-ait/FAIRPlus/tree/master/schemas/transcriptomics_schema">transcriptomics metadata schema</a>, which not only follows the MINSEQE guidelines, but also compatible with <a href="https://github.com/ebi-ait/FAIRPlus/blob/master/schemas/EBI_database_schemas/USI_schemas/USI_root_schema.json">EBI DSP schema</a>. Example data and submission guide will be added soon.</p>
<p>For future metadata submission, we created <a href="https://github.com/FuqiX/easyUSI">easyUSI</a>, an EBI DSP submission tool to simplify the sample metadata submission process and make the submission scalable.</p>
<h2 id="Data-acquisition">Data acquisition<a class="anchor-link" href="#Data-acquisition"> </a></h2><ul>
<li><p>identifiers</p>
</li>
<li><p>initial, local data QC</p>
</li>
</ul>
<h2 id="Data-transfer">Data transfer<a class="anchor-link" href="#Data-transfer"> </a></h2><ul>
<li><p>SFTP server</p>
</li>
<li><p>OneDrive</p>
</li>
<li><p>Amazon S3 Buckets</p>
</li>
</ul>
<h2 id="Data-processing-&amp;-analysis">Data processing &amp; analysis<a class="anchor-link" href="#Data-processing-&amp;-analysis"> </a></h2><ul>
<li><p>raw data → processed data → interpreted data</p>
</li>
<li><p>workflow / pipeline annotation</p>
</li>
<li><p>data provenance</p>
</li>
</ul>
<h2 id="Data-annotation">Data annotation<a class="anchor-link" href="#Data-annotation"> </a></h2><ul>
<li><p>According to our Data Management Plan (DMP)</p>
</li>
<li><p>Word document (as temporary solution until we finished schema definition for the RESOLUTE repository)</p>
<ul>
<li>JSON schemas</li>
</ul>
</li>
<li><p>FAIR aspects</p>
</li>
</ul>
<h2 id="Data-release">Data release<a class="anchor-link" href="#Data-release"> </a></h2><ul>
<li><p>consortium internal QC and discussion (RESOLUTE’s <em>"data annotation &amp; release workshop"</em>)</p>
</li>
<li><p>decision on license</p>
</li>
</ul>
<h2 id="Data-submission:">Data submission:<a class="anchor-link" href="#Data-submission:"> </a></h2><ul>
<li><p>raw data: SRA [Transcriptomics], ProteomeXchange [Proteomics]</p>
</li>
<li><p>interpreted data: ?</p>
</li>
<li><p>source annotation: BioSamples</p>
<ul>
<li><p>automatic via SRA → ENA</p>
</li>
<li><p>extension?</p>
</li>
</ul>
</li>
</ul>

</div>
</div>
</div>
</div>

 


    </main>
    