
<!DOCTYPE html>

<html>
  <head>
    <meta charset="utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Tools for Data Extraction, transformation, and loading</title>
    
  <link rel="stylesheet" href="../../../_static/css/index.f658d18f9b420779cfdf24aa0a7e2d77.css">

    
  <link rel="stylesheet"
    href="../../../_static/vendor/fontawesome/5.13.0/css/all.min.css">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-solid-900.woff2">
  <link rel="preload" as="font" type="font/woff2" crossorigin
    href="../../../_static/vendor/fontawesome/5.13.0/webfonts/fa-brands-400.woff2">

    
      
  <link rel="stylesheet"
    href="../../../_static/vendor/open-sans_all/1.44.1/index.css">
  <link rel="stylesheet"
    href="../../../_static/vendor/lato_latin-ext/1.44.1/index.css">

    
    <link rel="stylesheet" href="../../../_static/pygments.css" type="text/css" />
    <link rel="stylesheet" href="../../../_static/sphinx-book-theme.e7340bb3dbd8dde6db86f25597f54a1b.css" type="text/css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/togglebutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/copybutton.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/mystnb.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/sphinx-thebe.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/custom.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-main.c949a650a448cc0ae9fd3441c0e17fb0.css" />
    <link rel="stylesheet" type="text/css" href="../../../_static/panels-variables.06eb56fa6e07937060861dad626602ad.css" />
    
  <link rel="preload" as="script" href="../../../_static/js/index.d3f166471bb80abb5163.js">

    <script id="documentation_options" data-url_root="../../../" src="../../../_static/documentation_options.js"></script>
    <script src="../../../_static/jquery.js"></script>
    <script src="../../../_static/underscore.js"></script>
    <script src="../../../_static/doctools.js"></script>
    <script src="../../../_static/togglebutton.js"></script>
    <script src="../../../_static/clipboard.min.js"></script>
    <script src="../../../_static/copybutton.js"></script>
    <script src="../../../_static/mermaid.js"></script>
    <script src="../../../_static/mermaid.min.js"></script>
    <script src="../../../_static/mermaid.core.js"></script>
    <script async="async" kind="hypothesis" src="https://hypothes.is/embed.js"></script>
    <script kind="utterances">

    var commentsRunWhenDOMLoaded = cb => {
    if (document.readyState != 'loading') {
        cb()
    } else if (document.addEventListener) {
        document.addEventListener('DOMContentLoaded', cb)
    } else {
        document.attachEvent('onreadystatechange', function() {
        if (document.readyState == 'complete') cb()
        })
    }
}

var addUtterances = () => {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src = "https://utteranc.es/client.js";
    script.async = "async";

    script.setAttribute("repo", "FAIRplus/the-fair-cookbook");
    script.setAttribute("issue-term", "pathname");
    script.setAttribute("theme", "github-light");
    script.setAttribute("label", "ðŸ’¬ comment");
    script.setAttribute("crossorigin", "anonymous");

    sections = document.querySelectorAll("div.section");
    if (sections !== null) {
        section = sections[sections.length-1];
        section.appendChild(script);
    }
}
commentsRunWhenDOMLoaded(addUtterances);
</script>
    <script >var togglebuttonSelector = '.toggle, .admonition.dropdown, .tag_hide_input div.cell_input, .tag_hide-input div.cell_input, .tag_hide_output div.cell_output, .tag_hide-output div.cell_output, .tag_hide_cell.cell, .tag_hide-cell.cell';</script>
    <script src="../../../_static/sphinx-book-theme.7d483ff0a819d6edff12ce0b1ead3928.js"></script>
    <script async="async" src="https://unpkg.com/thebelab@latest/lib/index.js"></script>
    <script >
        const thebe_selector = ".thebe"
        const thebe_selector_input = "pre"
        const thebe_selector_output = ".output"
    </script>
    <script async="async" src="../../../_static/sphinx-thebe.js"></script>
    <link rel="canonical" href="https://fairplus.github.io/the-fair-cookbook/docs/content/recipes/ETL_tools.html" />
    <link rel="shortcut icon" href="../../../_static/favicon.ico"/>
    <link rel="index" title="Index" href="../../../genindex.html" />
    <link rel="search" title="Search" href="../../../search.html" />

    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <meta name="docsearch:language" content="en" />


<!-- Opengraph tags -->
<meta property="og:url"         content="https://fairplus.github.io/the-fair-cookbook/docs/content/recipes/ETL_tools.html" />
<meta property="og:type"        content="article" />
<meta property="og:title"       content="Tools for Data Extraction, transformation, and loading" />
<meta property="og:description" content="Tools for Data Extraction, transformation, and loading  :class: tip      - Data Curator     - Data Manager     - Data Scientist  For the `Recipe Type` section, " />
<meta property="og:image"       content="https://fairplus.github.io/the-fair-cookbook/_static/cookbook-logo-small.png" />

<meta name="twitter:card" content="summary" />


  </head>
  <body data-spy="scroll" data-target="#bd-toc-nav" data-offset="80">
    

    <div class="container-xl">
      <div class="row">
          
<div class="col-12 col-md-3 bd-sidebar site-navigation show" id="site-navigation">
    
        <div class="navbar-brand-box">
<a class="navbar-brand text-wrap" href="../../../index.html">
  
  <img src="../../../_static/cookbook-logo-small.png" class="logo" alt="logo">
  
  
</a>
</div><form class="bd-search d-flex align-items-center" action="../../../search.html" method="get">
  <i class="icon fas fa-search"></i>
  <input type="search" class="form-control" name="q" id="search-input" placeholder="Search this book..." aria-label="Search this book..." autocomplete="off" >
</form>
<nav class="bd-links" id="bd-docs-nav" aria-label="Main navigation">
    <ul class="nav sidenav_l1">
 <li class="toctree-l1">
  <a class="reference internal" href="../../../content/home.html">
   FAIR Cookbook
  </a>
 </li>
</ul>
<ul class="nav sidenav_l1">
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/introduction.html">
   Introduction
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/introduction/the-turing-way.html">
     1. The Turing Way Book of Data Science
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/introduction/FAIR-cookbook-audience.html">
     2. Who should read the FAIR Cookbook?
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/introduction/brief-FAIR-principles.html">
     3. What are the FAIR Principles?
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/introduction/what-is-IMI.html">
     4. What is IMI?
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/introduction/FAIR-in-IMI-context.html">
     5. FAIR in the context of IMI
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/introduction/IMI-Data-Access-Request-Process.html">
     6. Engaging with IMI Projects
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/introduction/FAIRplus-values.html">
     7. The Ethical Values behind FAIR and the FAIRPlus project
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/introduction/data-sharing-models.html">
     8. High level data sharing models
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/assessing-fairness.html">
   Assessing FAIR
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/assessing-fairness/cmmi/Data-Management-Maturity-Model.html">
     1. Data Management Maturity Model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/assessing-fairness/cmmi/Datasets-Maturity-Model.html">
     2. FAIRplus Datasets Maturity Model
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/assessing-fairness/cmmi/Data-Usage-Areas.html">
     3. Data Usage Areas
    </a>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/assessing-fairness/fair-indicators.html">
     4. FAIR Indicators
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/assessing-fairness/cmmi/RDA-Indicators.html">
       4.4.1. RDA Indicators
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/assessing-fairness/cmmi/FAIRPlus-Indicators.html">
       4.4.2. IMI FAIRPlus Indicators
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/assessing-fairness/manual-fair-assessment.html">
     5. FAIR assessment by hand
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/assessing-fairness/cmmi/evaluation-method.html">
       5.4.1. Evaluation Method
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/assessing-fairness/cmmi/rda-evaluation.html">
       5.4.2. RDA Evaluation Methods
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/assessing-fairness/automatic-fair-assessment.html">
     6. FAIR assessment by software
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/assessing-fairness/fair-assessment-fairevaluator.html">
       6.4.1. Assessing FAIRness with FAIREvaluator
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/assessing-fairness/fair-assessment-fairshake.html">
       6.4.2. FAIR assessment with FAIRShake
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/infrastructure.html">
   Infrastructure for FAIR
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/infrastructure/id-resolution.html">
     1. Identifier resolution services
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/infrastructure/data-catalog.html">
     2. Building a Dataset Catalogue
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/infrastructure/data-catalog-deployment.html">
     3. Deploying the IMI data catalogue
    </a>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/infrastructure/vocabularies.html">
     4. Vocabularies and Terminology Infrastructure
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/infrastructure/vocabularies/technical-and-architectural-selection-criteria-of-ontology-lookup-services.html">
       4.1. Technical &amp; architectural selection criteria of ontology lookup services
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/infrastructure/vocabularies/Selecting_and_using_ontology_lookup_services.html">
       4.2. Selection and exploitation of Ontology Lookup Services
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/infrastructure/vocabularies/UC3_R13_local_ontology_services.html">
       4.3. How to deploy a terminology service - EBI Ontology Lookup Service
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/findability.html">
   Improving Findability
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/findability/identifiers.html">
     1. How to generate globally unique, resolvable and persistent identifiers
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/infrastructure/chemical-identities.html">
     2. Chemical Identity - InChI and SMILES
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/findability/identifiers-minids.html">
     3. Identifier Minting Service with Minid Client
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/findability/zenodo-deposition.html">
     4. Deposition to Zenodo Archive
    </a>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/findability/seo.html">
     5. Findability: Search Engine Optimization
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/findability/seo/bioschemas-data-page.html">
       5.8.1. Findability: Data Page Markup with Bioschemas
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/findability/seo/bioschemas-dataset.html">
       5.8.2. Findability: Dataset Markup
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/findability/seo/bioschemas-datacatalog.html">
       5.8.3. Findability: Data Catalog Markup
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/accessibility.html">
   Improving Accessibility
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/accessibility/anonymizing-patient-data.html">
     1. Anonymizing Patient Data
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/accessibility/how-to-use-sftp-for-transfer.html">
     2. How to use SFTP to transfer data files between institutions.
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/accessibility/aspera.html">
     3. How to download files with Aspera
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/interoperability.html">
   Improving Interoperability
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/interoperability/identifier-mapping.html">
     How to interlink data from different sources?
    </a>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/interoperability/bridgedb-recipe.html">
     Identifier mapping with BridgeDB
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/interoperability/bridgedb/bridgedb-example.html">
       Identifier mapping using BridgeDb WS
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/interoperability/selecting-ontologies.html">
     Which vocabulary to use?
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/interoperability/ontology-new-term-request-recipe.html">
     Requesting terms addition to terminology artefacts
    </a>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/interoperability/ontology-robot-recipe.html">
     Building an application ontology with Robot
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/interoperability/ontology-robot/MSIO-robot-build-process.html">
       MSIO, an application ontology for Metabolomics
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/interoperability/creating-minimal-metadata-profiles.html">
     Creating a Metadata Profile
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/interoperability/transcriptomics-metadata.html">
       Establishing a Metadata Profile for Transcriptomics
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/interoperability/covid19-sample-metadata-profile-shex-use-case.html">
       Metadata Profile Validation in RDF with Shape Expression: The Covid-19 sample metadata use case
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/interoperability/from-proprietary-to-open-standard-mzml-exemplar.html">
     From proprietary to open standard data format
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/interoperability/fastq-file-format-validators.html">
     File format validation, an example with FASTQ files
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/reusability.html">
   Improving Reusability
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/reusability/ATI-licensing.html">
     1. Licensing
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/reusability/ATI_licensing_software.html">
     2. Software Licenses
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/reusability/ATI_licensing_data.html">
     3. Data Licenses
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/reusability/expressing-data-use.html">
     4. Patterns and practice for expressing allowed data usage
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/reusability/provenance.html">
     5. Provenance Information
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/reusability/community-standards.html">
     6. Compliance with community annotation standards
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/applied-examples.html">
   Making IMI data FAIR
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2 collapsible-parent">
    <a class="reference internal" href="../../../content/recipes/applied-examples/fair-data-matrix-recipe.html">
     Making a Data Matrix FAIR
    </a>
    <ul class="collapse-ul">
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/applied-examples/fair-data-matrix/0-rose-metabolites-Python-data-handling.html">
       Bibliographic Reference:
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/applied-examples/fair-data-matrix/1-rose-metabolites-Python-analysis.html">
       Exploring and comparing Rose Scent profiles stored as tabular data packages with plotnine, a python port of R ggplot2
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/applied-examples/fair-data-matrix/2-rose-metabolites-Python-RDF-querying-analysis.html">
       Exploring a rose garden with SPARQL
      </a>
     </li>
     <li class="toctree-l3">
      <a class="reference internal" href="../../../content/recipes/applied-examples/fair-data-matrix/3-rose-metabolites-R-analysis.html">
       Exploring and comparing Rose Scent profiles stored as tabular data packages with ggplot2 R library
      </a>
     </li>
    </ul>
    <i class="fas fa-chevron-down">
    </i>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/applied-examples/etox_raw/FAIRplus_Recipe_Ontology_mapping_the_eTox_dataset_scenario.html">
     Data normalization and integration through entity linking
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/applied-examples/nd4bb_raw/FAIRification_CookBook_Recipe1_V02.html">
     ND4BB dataset FAIRification recipe
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/applied-examples/oncotrack_raw/rawRecipe-oncotrack.html">
     Oncotrack FAIRification recipe
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/applied-examples/resolute_raw/rawRecipe_resolute.html">
     RESOLUTE FAIRification Recipe
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1">
  <a class="reference internal" href="../../../content/recipes/glossary-fair-terms.html">
   Glossary of FAIR things
  </a>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/help.html">
   Help
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/help/tips-tricks.html">
     1. Recipe Tips and Tricks
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/help/recipe-template.html">
     2. Recipe Template
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/help/how-to-create-recipe-with-gdoc.html">
     3. Using GoogleDocs to contribute a FAIR Cookbook Recipe
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/help/how-to-create-recipe-with-hackmd.html">
     4. How to use HackMD to write a FAIR Cookbook Recipe
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/help/how-to-create-recipe-with-git.html">
     5. How to create a FAIRPlus Coobook Recipe ?
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
 <li class="toctree-l1 collapsible-parent">
  <a class="reference internal" href="../../../content/recipes/boilerplate.html">
   Miscellaneous
  </a>
  <ul class="collapse-ul">
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/boilerplate/license.html">
     1. License for this book
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/boilerplate/people.html">
     2. FAIR Cookbook Steering Council and Institutional Partners
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/boilerplate/governance.html">
     3. FAIR Cookbook Governance
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/boilerplate/code_of_conduct.html">
     4. FAIR Cookbook Code of Conduct
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/boilerplate/project_information.html">
     5. FAIR CookBook BoilerPlate
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/boilerplate/roadmap.html">
     6. FAIR cookbook development roadmap
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/boilerplate/contributing.html">
     7. Contributing to the FAIR cookbook
    </a>
   </li>
   <li class="toctree-l2">
    <a class="reference internal" href="../../../content/recipes/boilerplate/disclaimer.html">
     8. Disclaimer
    </a>
   </li>
  </ul>
  <i class="fas fa-chevron-down">
  </i>
 </li>
</ul>

</nav> <!-- To handle the deprecated key -->

<div class="navbar_extra_footer">
  Powered by <a href="https://jupyterbook.org">Jupyter Book</a>
</div>

</div>


          


          
<main class="col py-md-3 pl-md-4 bd-content overflow-auto" role="main">
    
    <div class="topbar container-xl fixed-top">
    <div class="topbar-contents row">
        <div class="col-12 col-md-3 bd-topbar-whitespace site-navigation show"></div>
        <div class="col pl-md-4 topbar-main">
            
            <button id="navbar-toggler" class="navbar-toggler ml-0" type="button" data-toggle="collapse"
                data-toggle="tooltip" data-placement="bottom" data-target=".site-navigation" aria-controls="navbar-menu"
                aria-expanded="true" aria-label="Toggle navigation" aria-controls="site-navigation"
                title="Toggle navigation" data-toggle="tooltip" data-placement="left">
                <i class="fas fa-bars"></i>
                <i class="fas fa-arrow-left"></i>
                <i class="fas fa-arrow-up"></i>
            </button>
            
            
<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn" aria-label="Download this page"><i
            class="fas fa-download"></i></button>

    <div class="dropdown-buttons">
        <!-- ipynb file if we had a myst markdown file -->
        
        <!-- Download raw file -->
        <a class="dropdown-buttons" href="../../../_sources/docs/content/recipes/ETL_tools.md"><button type="button"
                class="btn btn-secondary topbarbtn" title="Download source file" data-toggle="tooltip"
                data-placement="left">.md</button></a>
        <!-- Download PDF via print -->
        <button type="button" id="download-print" class="btn btn-secondary topbarbtn" title="Print to PDF"
            onClick="window.print()" data-toggle="tooltip" data-placement="left">.pdf</button>
    </div>
</div>

            <!-- Source interaction buttons -->

<div class="dropdown-buttons-trigger">
    <button id="dropdown-buttons-trigger" class="btn btn-secondary topbarbtn"
        aria-label="Connect with source repository"><i class="fab fa-github"></i></button>
    <div class="dropdown-buttons sourcebuttons">
        <a class="repository-button"
            href="https://github.com/FAIRplus/the-fair-cookbook"><button type="button" class="btn btn-secondary topbarbtn"
                data-toggle="tooltip" data-placement="left" title="Source repository"><i
                    class="fab fa-github"></i>repository</button></a>
        <a class="issues-button"
            href="https://github.com/FAIRplus/the-fair-cookbook/issues/new?title=Issue%20on%20page%20%2Fdocs/content/recipes/ETL_tools.html&body=Your%20issue%20content%20here."><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Open an issue"><i class="fas fa-lightbulb"></i>open issue</button></a>
        <a class="edit-button" href="https://github.com/FAIRplus/the-fair-cookbook/edit/migrating/docs/content/recipes/ETL_tools.md"><button
                type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip" data-placement="left"
                title="Edit this page"><i class="fas fa-pencil-alt"></i>suggest edit</button></a>
    </div>
</div>


            <!-- Full screen (wrap in <a> to have style consistency -->
            <a class="full-screen-button"><button type="button" class="btn btn-secondary topbarbtn" data-toggle="tooltip"
                    data-placement="bottom" onclick="toggleFullScreen()" aria-label="Fullscreen mode"
                    title="Fullscreen mode"><i
                        class="fas fa-expand"></i></button></a>

            <!-- Launch buttons -->

        </div>

        <!-- Table of contents -->
        <div class="d-none d-md-block col-md-2 bd-toc show">
            
        <div class="tocsection onthispage pt-5 pb-3">
            <i class="fas fa-list"></i>
            Contents
        </div>
        <nav id="bd-toc-nav">
            <ul class="nav section-nav flex-column">
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#">
   Tools for Data Extraction, transformation, and loading
  </a>
  <ul class="nav section-nav flex-column">
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#main-objectives">
     Main Objectives
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#graphical-overview-of-the-fairification-recipe-objectives">
     Graphical Overview of the FAIRification Recipe Objectives
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#requirements">
     Requirements
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#capability-maturity-table">
     Capability &amp; Maturity Table
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#fairification-objectives-inputs-and-outputs">
     FAIRification Objectives, Inputs and Outputs
    </a>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#operations">
     Operations
    </a>
    <ul class="nav section-nav flex-column">
     <li class="toc-h3 nav-item toc-entry">
      <a class="reference internal nav-link" href="#example-use-case">
       Example use case:
      </a>
     </li>
    </ul>
   </li>
   <li class="toc-h2 nav-item toc-entry">
    <a class="reference internal nav-link" href="#authors">
     Authors
    </a>
   </li>
  </ul>
 </li>
 <li class="toc-h1 nav-item toc-entry">
  <a class="reference internal nav-link" href="#references">
   References:
  </a>
 </li>
</ul>

        </nav>
        
        </div>
    </div>
</div>
    <div id="main-content" class="row">
        <div class="col-12 col-md-9 pl-md-3 pr-md-0">
        
              <div>
                
  <div class="section" id="tools-for-data-extraction-transformation-and-loading">
<h1>Tools for Data Extraction, transformation, and loading<a class="headerlink" href="#tools-for-data-extraction-transformation-and-loading" title="Permalink to this headline">Â¶</a></h1>
<div class="highlight-default notranslate"><div class="highlight"><pre><span></span>:class: tip

    - Data Curator
    - Data Manager
    - Data Scientist

For the `Recipe Type` section, the allowed values should be picked from the following list:
    - Review 
</pre></div>
</div>
<div class="section" id="main-objectives">
<h2>Main Objectives<a class="headerlink" href="#main-objectives" title="Permalink to this headline">Â¶</a></h2>
<p>This recipe identifies tools for data extraction, transformation, and loading (ETL). ETL is the process of collecting data from one source to a designated system in which the data is represented differently.[1] One common use case in biological science is to build a scalable and portable ETL system to extract data from different sources, transform data into a cohesive dataset, and load data to an internal or public database to support data exchange. This recipe aims to serve as a start point for designing ETL workflows, instead of provide a comprehensive list covering all available tools.</p>
<blockquote>
<div><p>The lists of tools are generated either by manual curation, which reflects what is being used in the industry, or automatally discovered from the bio.tools repository.</p>
<p>:bulb: Contents in this table are generated in March 2021. For updated contents, please check the FAIR tooling repository. You can provide feedback or report issues in the comments section.</p>
</div></blockquote>
</div>
<div class="section" id="graphical-overview-of-the-fairification-recipe-objectives">
<h2>Graphical Overview of the FAIRification Recipe Objectives<a class="headerlink" href="#graphical-overview-of-the-fairification-recipe-objectives" title="Permalink to this headline">Â¶</a></h2>
<p>The figure below shows different ETL-related operations and their relationships, together with related tools and recipes.</p>
<div class="highlight-mermaid notranslate"><div class="highlight"><pre><span></span>graph LR
    subgraph Data extraction
        
        A1[Data query snd retrieval]
    end
    
    subgraph Data transform
        B1[Data annotation]
        B2[Data validation]
        B3[Data aggregation]
    end
    
    C[Data deposition]
    
    A1--&gt;B1
    B1--&gt;B2
    B2--&gt;B3
    B3--&gt;C
    
</pre></div>
</div>
<p>The table below is an overview of ETL tools identified. Details of each tools are also provided in this doc. For ETL tools for RDF data model, please check recipe <span class="xref std std-ref">An inventory of tools for converting your data to RDF</span>.</p>
<table>
  <tr>
   <td><strong><a href=#Data-query-and-retrieval>Data query and retrieval</a></strong>
   </td>
   <td><strong><a href=#Data-transformation>Data transformation</a></strong>
   </td>
   <td><strong><a href=#Data-deposition>Data deposition</a></strong>
   </td>
  </tr>
  <tr>
   <td><a href="https://www.tamr.com/">TAMR</a>
   </td>
   <td><a href="https://www.cdisc.org/standards/foundational/sdtm">SDTM</a>
   </td>
   <td><a href="https://www.project-redcap.org/">REDCap</a>
   </td>
  </tr>
  <tr>
   <td><a href="https://github.com/google-research/bert">Bert</a>
   </td>
   <td><a href="https://www.trifacta.com">TriFacta</a>
   </td>
   <td><a href="https://github.com/transmart">TransMART</a>
   </td>
  </tr>
  <tr>
   <td><a href="https://www.scibite.com/platform/termite/">Termite</a>
   </td>
   <td><a href="https://www.ohdsi.org/data-standardization/the-common-data-model/">OMOP</a>
   </td>
   <td><a href="https://www.stardog.com/">Stardog</a>
   </td>
  </tr>
  <tr>
   <td><a href="https://www.oracle.com/uk/industries/life-sciences/clinical-research/">Oracle clinical</a>
   </td>
   <td><a href="https://www.collibra.com/">Collibra</a>
   </td>
   <td><a href="https://www.postgresql.org/">Postgresql</a>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/geoboost2">GeoBoost2</a>
   </td>
   <td><a href="https://hadoop.apache.org/">Apache Hadoop</a>
   </td>
   <td><a href="https://bio.tools/chemotion">Chemotion</a>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/martview">MartView</a>
   </td>
   <td><a href="https://www.talend.com/products/integrate-data/">Talend</a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/CLOBNET">CLOBNET</a>
   </td>
   <td><a href="https://www.informatica.com/">Informatica</a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/PDBUSQLExtractor">PDBUSQLExtractor</a>
   </td>
   <td><a href="https://github.com/FAIRDataTeam/OpenRefine-metadata-extension">OpenRefine-metadata-extension</a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/repo">repo</a>
   </td>
   <td><a href="https://bio.tools/discover">DISQOVER</a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/snpator">SNPator</a>
   </td>
   <td><a href="https://bio.tools/Query_Tabular">Query Tabular</a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/PyGMQL">PyGMQL</a>
   </td>
   <td><a href="https://bio.tools/ms-data-core-api">ms-data-core-api</a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/plantpis">PlantPIs</a>
   </td>
   <td>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/baget">BAGET</a>
   </td>
   <td><a href=""></a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/ncbi_resources">NCBI Resources</a>
   </td>
   <td><a href=""></a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/cpad">CPAD</a>
   </td>
   <td><a href=""></a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/GenomicsDB">GenomicsDB</a>
   </td>
   <td><a href=""></a>
   </td>
   <td>
   </td>
  </tr>
  <tr>
   <td><a href="https://bio.tools/epivizFileServer">Epiviz File Server</a></a>
   </td>
   <td><a href=""></a>
   </td>
   <td>
   </td>
  </tr>
</table>
>Disclaimer: The tools list below aims to provide a idea of what is there on the market. It's not a formal recommendation. if you think there are tools that need to be updated, please contact us via the github issue.
</div>
<div class="section" id="requirements">
<h2>Requirements<a class="headerlink" href="#requirements" title="Permalink to this headline">Â¶</a></h2>
<ul class="simple">
<li><p>knowledge requirement:</p>
<ul>
<li><p>Familiar with general data ETL workflow.</p></li>
</ul>
</li>
</ul>
</div>
<hr class="docutils" />
<div class="section" id="capability-maturity-table">
<h2>Capability &amp; Maturity Table<a class="headerlink" href="#capability-maturity-table" title="Permalink to this headline">Â¶</a></h2>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="text-align:left head"><p>Capability</p></th>
<th class="text-align:left head"><p>Initial Maturity Level</p></th>
<th class="text-align:left head"><p>Final Maturity Level</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-align:left"><p>Interoperability</p></td>
<td class="text-align:left"><p>minimal</p></td>
<td class="text-align:left"><p>automated</p></td>
</tr>
</tbody>
</table>
</div>
<hr class="docutils" />
<div class="section" id="fairification-objectives-inputs-and-outputs">
<h2>FAIRification Objectives, Inputs and Outputs<a class="headerlink" href="#fairification-objectives-inputs-and-outputs" title="Permalink to this headline">Â¶</a></h2>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="text-align:left head"><p>Actions.Objectives.Tasks</p></th>
<th class="text-align:left head"><p>Input</p></th>
<th class="text-align:left head"><p>Output</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td class="text-align:left"><p><a class="reference external" href="http://edamontology.org/operation_2422">Data retrival</a></p></td>
<td class="text-align:left"><p></p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p><a class="reference external" href="http://edamontology.org/operation_2428">validation</a></p></td>
<td class="text-align:left"><p></p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-even"><td class="text-align:left"><p><a class="reference external" href="http://edamontology.org/operation_3431">data deposition</a></p></td>
<td class="text-align:left"><p></p></td>
<td class="text-align:left"><p></p></td>
</tr>
<tr class="row-odd"><td class="text-align:left"><p><a class="reference external" href="http://edamontology.org/topic_0219">Data submission, annotation and curation</a></p></td>
<td class="text-align:left"><p></p></td>
<td class="text-align:left"><p></p></td>
</tr>
</tbody>
</table>
</div>
<hr class="docutils" />
<div class="section" id="operations">
<h2>Operations<a class="headerlink" href="#operations" title="Permalink to this headline">Â¶</a></h2>
<h3 id=Data-query-and-retrieval>Data query and retrieval</h3>
The process of extracting structured or unstructured data from different sources. The process of identifying and obtaining data from data management systems.
<p><strong>Manually curated tools</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Tool</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>License</p></th>
<th class="head"><p>Topics</p></th>
<th class="head"><p>Resource type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://www.tamr.com/">TAMR</a></p></td>
<td><p>A cloud-native data mastering solution (cloud MDM) accelerate analytics through machine learning (ML), available on Google Cloud, Azure and AWS.</p></td>
<td><p>Commercial license</p></td>
<td><p>information integration, data unification</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://github.com/google-research/bert">Bert</a></p></td>
<td><p>Bidirectional Encoder Representations from Transformers (BERT) is a Transformer-based machine learning technique for natural language processing (NLP) pre-training developed by Google, aiming to extractcontext-sensitive features from an input text.</p></td>
<td><p>Apache License 2.0</p></td>
<td><p>relationship extraction</p></td>
<td><p>Data model</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://www.scibite.com/platform/termite/">Termite</a></p></td>
<td><p>TERMite (TERM identification, tagging &amp; extraction) is a fast named entity recognition (NER) and extraction engine for semantic analytics.</p></td>
<td><p>Commercial license</p></td>
<td><p>information extraction</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://www.oracle.com/uk/industries/life-sciences/clinical-research/">Oracle clinical</a></p></td>
<td><p>a single application and infrastructure for electronic data capture and clinical data management, while leveraging the renowned Oracle database. Oracle Clinical enables management of all clinical trial data in a single system, improving accuracy, visibility, and data integrity.</p></td>
<td><p>Commercial license</p></td>
<td><p>Clinical data</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p><strong>Tools discovered from <a class="reference external" href="https://bio.tools">Bio.Tools</a>.</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Tool</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>License</p></th>
<th class="head"><p>Topics</p></th>
<th class="head"><p>Resource type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/geoboost2">GeoBoost2</a></p></td>
<td><p>A natural languageprocessing pipeline for GenBank metadata enrichment for virus phylogeography.</p></td>
<td><p>Apache-2.0 License</p></td>
<td><p>Biotechnology, Natural language processing, Workflows, Public health and epidemiology, Infectiousdisease</p></td>
<td><p>Web application</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://bio.tools/martview">MartView</a></p></td>
<td><p>Tool for data retrieval and data mining that integrates data from Ensembl. Through the web interface it allows you to apply a series of filters to create custom datasets which can be converted to several useful output formats.</p></td>
<td><p>N/A</p></td>
<td><p>Pathology, Data integration and warehousing, Database management, Molecular interactions, pathways and networks, Gene transcripts</p></td>
<td><p>Web Application</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/CLOBNET">CLOBNET</a></p></td>
<td><p>Cloud-based machine learning system (CLOBNET) that is an open-source, lean infrastructure for electronic health record (EHR) data integration and is capable of extract, transform, and load (ETL) processing</p></td>
<td><p>BSD-2-Clause</p></td>
<td><p>Machine learning, Medicines research and development</p></td>
<td><p>Web application (for internal use)</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://bio.tools/PDBUSQLExtractor">PDBUSQLExtractor</a></p></td>
<td><p>Scalable Extraction of Big Macromolecular Data in Azure Data Lake Environment.</p></td>
<td><p>GPL-3.0</p></td>
<td><p>Nucleic acid structure analysis</p></td>
<td><p>C# library</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/repo">repo</a></p></td>
<td><p>A data manager meant to avoid manual storage/retrieval of data to/from the file system. It builds one (or more) centralized repository where R objects are stored with rich annotations, including corresponding code chunks, and easily searched and retrieved.</p></td>
<td><p>GPL-3.0</p></td>
<td><p>Data submission, annotation and curation, Data management, Database management</p></td>
<td><p>R library</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://bio.tools/snpator">SNPator</a></p></td>
<td><p>Originally designed to help CeGen users to handle, retrieve, transform and analyze the genetic data generated by the genotyping facilities of the institution. However, it is also open to external users who may want to upload their own genotyped data in order to take advantage its data processing features. Users will be able to perform a set of operations which may range from very simple format transformations to some complex biostatistical calculations.</p></td>
<td><p>N/A</p></td>
<td><p>Bioinformatics, Data quality management, Genetic variation, GWAS study</p></td>
<td><p>Web Application</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/PyGMQL">PyGMQL</a></p></td>
<td><p>Scalable data extraction and analysis for heterogeneous genomic datasets. Based on GMQL.</p></td>
<td><p>Apache-2.0</p></td>
<td><p>Imaging, Workflows, Database management</p></td>
<td><p>Python Library</p></td>
</tr>
</tbody>
</table>
<p><strong>Data portals with extraction functionalities discovered from <a class="reference external" href="https://bio.tools">Bio.Tools</a>.</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Tool</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>License</p></th>
<th class="head"><p>Topics</p></th>
<th class="head"><p>Resource type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/plantpis">PlantPIs</a></p></td>
<td><p>Web querying system for a database collecting plant protease inhibitors data.</p></td>
<td><p>Plant biology, Gene and protein families</p></td>
<td><p>Database Portal</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://bio.tools/baget">BAGET</a></p></td>
<td><p>Web service designed to facilitate extraction of specific gene and protein sequences from completely determined prokaryotic genomes. Query results can be exported as a rich text format file for printing, archival or further analysis.</p></td>
<td><p>N/A</p></td>
<td><p>DNA, Sequencing, Model organisms, Sequence analysis</p></td>
<td><p>Web application</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/ncbi_resources">NCBI Resources</a></p></td>
<td><p>Analysis and retrieval resources for the data in GenBank and other biological data made available through the NCBI web site.</p></td>
<td><p>Open Access</p></td>
<td><p>Data mining, Data integration and warehousing, Database management, Molecular interactions, pathways and networks, Rare diseases</p></td>
<td><p>Web Application, Database portal</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://bio.tools/cpad">CPAD</a></p></td>
<td><p>Curated Protein Aggregation Database (CPAD) is an integrated database which contains information related to amyloidogenic proteins, Aggregation prone regions, aggregation kinetics and structure of proteins relevant to aggregation.</p></td>
<td><p>N/A</p></td>
<td><p>Proteomics, Data submission, annotation and curation, Proteins, Small molecules, Literature and language</p></td>
<td><p>Data Portal</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/GenomicsDB">GenomicsDB</a></p></td>
<td><p>Advancing clinical cohort selection with genomics analysis on a distributed platform. Highly performant data storage in C++ for importing, querying and transforming variant data with Java/Spark. Used in gatk4.</p></td>
<td><p>MIT License</p></td>
<td><p>Exome sequencing, Systems medicine, Personalised medicine, Complementary medicine, Biobank</p></td>
<td><p>Command-line</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://bio.tools/epivizFileServer">Epiviz File Server</a></p></td>
<td><p>A library to query and transform genomic data from indexed files</p></td>
<td><p>MIT License</p></td>
<td><p>Oncology,Workflows, DNA</p></td>
<td><p>Python Library</p></td>
</tr>
</tbody>
</table>
<h3 id=Data-transformation>Data transformation</h3>
### Data transformation
The process of converting the data format, structure, and value to meet specific data standards. Below are some common operations in data transformation.
- Data annotation
    The process of labeling data or metadata.
- Data validation
    The process of ensuring data has met data standards both in terms of data content and data format.
- Data aggregation
    The process of gathering data and presenting it, usually,  in a summarized format.
<p><strong>Manually curated tools</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Tool</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>License</p></th>
<th class="head"><p>Topics</p></th>
<th class="head"><p>Resource type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://www.cdisc.org/standards/foundational/sdtm">SDTM</a></p></td>
<td><p>SDTM provides a standard for organizing and formatting data to streamline processes in collection management analysis and reporting.</p></td>
<td><p>N/A</p></td>
<td><p>data aggregation, data warehousing, Standard for Exchange of Non-clinical Data</p></td>
<td><p>Data standard</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://www.trifacta.com">TriFacta</a></p></td>
<td><p>A data preparation tool for data quality improvement, data transformation, and building data pipelines.</p></td>
<td><p>Commercial license</p></td>
<td><p></p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://www.ohdsi.org/data-standardization/the-common-data-model/">OMOP</a></p></td>
<td><p>The OMOP Common Data Model allows for the systematic analysis of disparate observational databases.</p></td>
<td><p>N/A</p></td>
<td><p>Data transformation</p></td>
<td><p>Data model</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://www.collibra.com/">Collibra</a></p></td>
<td><p>an enterprise-oriented data governance platform that provides tools for data management and stewardship.</p></td>
<td><p>Commercial license</p></td>
<td><p>Data catalogue,data governance, data linkage</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://hadoop.apache.org/">Apache Hadoop</a></p></td>
<td><p>A framework that allows for the distributed processing of large data sets across clusters of computers using simple programming models. It is designed to scale up from single servers to thousands of machines</p></td>
<td><p><a class="reference external" href="http://www.apache.org/licenses/LICENSE-2.0">Apache License 2.0</a></p></td>
<td><p>Distributed computing</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://www.talend.com/products/integrate-data/">Talend</a></p></td>
<td><p>A unified approach that combines rapid data integration, transformation, and mapping with automated quality checks to ensure trustworthy data in every step.</p></td>
<td><p>Commercial license</p></td>
<td><p>Data integration</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://www.informatica.com/">Informatica</a></p></td>
<td><p>Connect &amp; fetch data from different heterogeneous source and processing of data.</p></td>
<td><p>Commercial license</p></td>
<td><p>Data integration</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p><strong>Tools discovered from <a class="reference external" href="https://bio.tools">Bio.Tools</a>.</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Tool</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>License</p></th>
<th class="head"><p>Topics</p></th>
<th class="head"><p>Resource type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://github.com/FAIRDataTeam/OpenRefine-metadata-extension">OpenRefine-metadata-extension</a></p></td>
<td><p>Enables a post-hoc FAIRification workflow: load an existing dataset, perform data wrangling tasks, add FAIR attributes to the data, generate a linked data version of the data and, finally, push the result to an online FAIR data infrastructure to make it accessible and discoverable.</p></td>
<td><p>MIT License</p></td>
<td><p>Data submission, Annotation and curation, Data identity and mapping</p></td>
<td><p>Extension</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://bio.tools/discover">DISQOVER</a></p></td>
<td><p>Data integration platform for public, licensed and internal data. The Data Ingestion Engine enables transforming data into Linked Data which can be searched, navigated and analysed via the user interface and the API.</p></td>
<td><p>N/A</p></td>
<td><p>Biomedical science, Omics, Biology, Chemistry, Medicine</p></td>
<td><p>â€“</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/pga">PGA</a></p></td>
<td><p>Construction of customized protein databases based on RNA-Seq data with/without genome guided, database searching, post-processing and report generation.</p></td>
<td><p>GPL-2.0</p></td>
<td><p>Proteomics, Proteomics experiment, Data submission, annotation and curation, Data integration and warehousing, Database management, Data governance, RNA-seq</p></td>
<td><p>R library</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://bio.tools/Query_Tabular">Query Tabular</a></p></td>
<td><p>Galaxy-based tool which manipulates tabular files. Query Tabular automatically creates a SQLite database directly from a tabular file within a Galaxy workflow. The SQLite database can be saved to the Galaxy history, and further process to generate tabular outputs containing desired information and formatting.</p></td>
<td><p>CC-BY-4.0</p></td>
<td><p>Bioinformatics,Workflows</p></td>
<td><p>Galaxy-based</p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/ms-data-core-api">ms-data-core-api</a></p></td>
<td><p>The primary purpose of ms-data-core-api library is to provide commonly used classes and Object Model for Proteomics Experiments. You may also find it useful for your own computational proteomics projects.</p></td>
<td><p>Apache-2.0</p></td>
<td><p>Proteomics, Proteomics experiment, Software engineering</p></td>
<td><p>Java Library</p></td>
</tr>
</tbody>
</table>
<h3 id=Data-deposition>Data deposition</h3>
<p>The process of loading data to hosting (end) destinations, such as public data archives and data warehouses.</p>
<p><strong>Manually curated tools</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Tool</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>License</p></th>
<th class="head"><p>Topics</p></th>
<th class="head"><p>Resource type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://www.project-redcap.org/">REDCap</a></p></td>
<td><p>REDCap is a secure web application for building and managing online surveys and databases. While REDCap can be used to collect virtually any type of data in any environment</p></td>
<td><p><a class="reference external" href="https://projectredcap.org/partners/termsofuse/">REDCap License Terms</a></p></td>
<td><p>Cloud migration</p></td>
<td><p>Web application</p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://github.com/transmart">TransMART</a></p></td>
<td><p>An open-source data warehouse designed to store large amounts of clinical data from clinical trials, as well as data from basic research</p></td>
<td><p><a class="reference external" href="https://github.com/transmart/transmartApp/blob/master/LICENSE.TXT">GPL-3.0 License</a></p></td>
<td><p>Data storage, clinical data</p></td>
<td><p></p></td>
</tr>
<tr class="row-even"><td><p><a class="reference external" href="https://www.stardog.com/">Stardog</a></p></td>
<td><p>Triple Store Database, Provide an enterprise knowledge graph as FAIR+ data catalogue</p></td>
<td><p>Commercial license</p></td>
<td><p>Data integration,data catalog</p></td>
<td><p></p></td>
</tr>
<tr class="row-odd"><td><p><a class="reference external" href="https://www.postgresql.org/">Postgresql</a></p></td>
<td><p>A free and open-source relational database management system (RDBMS) emphasizing extensibility and SQL compliance.</p></td>
<td><p><a class="reference external" href="https://www.postgresql.org/about/licence/">The PostgreSQL Licence</a></p></td>
<td><p>Relational database</p></td>
<td><p></p></td>
</tr>
</tbody>
</table>
<p><strong>Tools discovered from <a class="reference external" href="https://bio.tools">Bio.Tools</a>.</strong></p>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Tool</p></th>
<th class="head"><p>Description</p></th>
<th class="head"><p>License</p></th>
<th class="head"><p>Topics</p></th>
<th class="head"><p>Resource type</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p><a class="reference external" href="https://bio.tools/chemotion">Chemotion</a></p></td>
<td><p>Repository for chemistry research data that provides solutions for current challenges to store research data in a feasible manner, allowing the conservation of domain specific information in a machine readable format.</p></td>
<td><p>CC-BY-4.0</p></td>
<td><p>Data submission, annotation and curation, Workflows, Chemistry</p></td>
<td><p>Data repository</p></td>
</tr>
</tbody>
</table>
<div class="section" id="example-use-case">
<h3>Example use case:<a class="headerlink" href="#example-use-case" title="Permalink to this headline">Â¶</a></h3>
<p>To show how these tools can be used in real life examples, please check the related recipes.</p>
<ol class="simple">
<li><p>OMOP ETL <span class="xref myst">link</span></p></li>
<li><p>RDF ETL <span class="xref myst">link</span></p></li>
<li><p>FASTQ file validation <span class="xref myst">link</span></p></li>
</ol>
</div>
</div>
<div class="section" id="authors">
<h2>Authors<a class="headerlink" href="#authors" title="Permalink to this headline">Â¶</a></h2>
<table class="colwidths-auto table">
<thead>
<tr class="row-odd"><th class="head"><p>Name</p></th>
<th class="head"><p>Affiliation</p></th>
<th class="head"><p>ORCID</p></th>
<th class="head"><p>Credit role</p></th>
</tr>
</thead>
<tbody>
<tr class="row-even"><td><p>Fuqi Xu</p></td>
<td><p>EMBL-EBI</p></td>
<td><p>0000-0002-5923-3859</p></td>
<td><p>Original draft</p></td>
</tr>
<tr class="row-odd"><td><p>Eva Martin</p></td>
<td><p>BSC</p></td>
<td><p>0000-0001-8324-2897</p></td>
<td><p>Original draft</p></td>
</tr>
<tr class="row-even"><td><p>Sukhi Singh</p></td>
<td><p></p></td>
<td><p></p></td>
<td><p>Tool curation</p></td>
</tr>
</tbody>
</table>
</div>
</div>
<div class="section" id="references">
<h1>References:<a class="headerlink" href="#references" title="Permalink to this headline">Â¶</a></h1>
<ol class="simple">
<li><p><a class="reference external" href="https://en.wikipedia.org/wiki/Extract,_transform,_load">https://en.wikipedia.org/wiki/Extract,_transform,_load</a></p></li>
</ol>
</div>

    <script type="text/x-thebe-config">
    {
        requestKernel: true,
        binderOptions: {
            repo: "binder-examples/jupyter-stacks-datascience",
            ref: "master",
        },
        codeMirrorConfig: {
            theme: "abcdef",
            mode: "python"
        },
        kernelOptions: {
            kernelName: "python3",
            path: "./docs/content/recipes"
        },
        predefinedOutput: true
    }
    </script>
    <script>kernelName = 'python3'</script>

              </div>
              
        
        <div class='prev-next-bottom'>
            

        </div>
        
        </div>
    </div>
    <footer class="footer mt-5 mt-md-0">
    <div class="container">
      <p>
        
          By The FAIRplus project members<br/>
        
            &copy; Copyright 2020 (CC-BY 4.0).<br/>
          <div class="extra_footer">
            <script>mermaid.init();</script> <div class="row_footer"> This project has received funding from the <a href="https://www.imi.europa.eu/">Innovative Medicines Initiative</a> 2 Joint Undertaking under grant agreement No 802750. This Joint Undertaking receives support from the European Unionâ€™s Horizon 2020 research and innovation programme and EFPIA Companies. This communication reflects the views of the authors and neither IMI nor the European Union, EFPIA or any Associated Partners are liable for any use that may be made of the information contained herein. </div>
<div class="row_footer"> This page was created by <a href="https://github.com/jupyter/jupyter-book/graphs/contributors">The Jupyter Book Community</a> </div>
<div class="row_footer"> <div class="column" style="center"><a href="https://europa.eu/"><img src="https://raw.githubusercontent.com/FAIRplus/the-fair-cookbook/migrating/_static/images/logo/eu-flag-recolored-fairplus.png" height="20" alt="EU flag" style="float:center""></a></div> <div class="column"><a href="https://www.imi.europa.eu/"><img src="https://raw.githubusercontent.com/FAIRplus/the-fair-cookbook/migrating/_static/images/logo/imi-logo-fairplus-v1.png" height="20" alt="IMI logo" style="float:center""></a></div> <div class="column"><a href="https://efpia.eu"><img src="https://raw.githubusercontent.com/FAIRplus/the-fair-cookbook/migrating/_static/images/logo/efpia-logo-fairplus2.png" height="20" alt="EFPIA logo" style="float:center""></a></div> <br/> </div>
          </div>
      </p>
    </div>
  </footer>
</main>


      </div>
    </div>

    
  <script src="../../../_static/js/index.d3f166471bb80abb5163.js"></script>


    
    <!-- Google Analytics -->
    <script>
      window.ga=window.ga||function(){(ga.q=ga.q||[]).push(arguments)};ga.l=+new Date;
      ga('create', 'UA-171697285-1', 'auto');
      ga('set', 'anonymizeIp', true);
      ga('send', 'pageview');
    </script>
    <script async src='https://www.google-analytics.com/analytics.js'></script>
    <!-- End Google Analytics -->
    
  </body>
</html>